async_simple/util/ThreadPool.h

```diff
@@ -25,372 +25,137 @@
-
-    int trylock() { return pthread_mutex_trylock(&_mutex); }
-
-    int unlock() { return pthread_mutex_unlock(&_mutex); }
-
-private:
-    ThreadMutex(const ThreadMutex &);
-    ThreadMutex &operator=(const ThreadMutex &);
-
-protected:
-    pthread_mutex_t _mutex;
-};
-
-class ProducerConsumerCond : public ThreadMutex {
-public:
-    ProducerConsumerCond() {
-        pthread_cond_init(&_producerCond, NULL);
-        pthread_cond_init(&_consumerCond, NULL);
-    }
-
-    ~ProducerConsumerCond() {
-        pthread_cond_destroy(&_producerCond);
-        pthread_cond_destroy(&_consumerCond);
-    }
-
-public:
-    int producerWait(int64_t usec = 0) { return wait(_producerCond, usec); }
-
-    int signalProducer() { return signal(_producerCond); }
-
-    int broadcastProducer() { return broadcast(_producerCond); }
-
-    int consumerWait(int64_t usec = 0) { return wait(_consumerCond, usec); }
-
-    int signalConsumer() { return signal(_consumerCond); }
-
-    int broadcastConsumer() { return broadcast(_consumerCond); }
-
-private:
-    int wait(pthread_cond_t &cond, int64_t usec) {
-        int ret = 0;
-        assert(usec == 0);
-        ret = pthread_cond_wait(&cond, &_mutex);
-        return ret;
-    }
-
-    static int signal(pthread_cond_t &cond) {
-        return pthread_cond_signal(&cond);
-    }
-
-    static int broadcast(pthread_cond_t &cond) {
-        return pthread_cond_broadcast(&cond);
-    }
-
-protected:
-    pthread_cond_t _producerCond;
-    pthread_cond_t _consumerCond;
-};
-
-class ScopedLock {
-private:
-    ThreadMutex &_lock;
-
-private:
-    ScopedLock(const ScopedLock &);
-    ScopedLock &operator=(const ScopedLock &);
-
-public:
-    explicit ScopedLock(ThreadMutex &lock) : _lock(lock) {
-        int ret = _lock.lock();
-        assert(ret == 0);
-        (void)ret;
-    }
-
-    ~ScopedLock() {
-        int ret = _lock.unlock();
-        assert(ret == 0);
-        (void)ret;
-    }
-};
-
-class WorkItem {
-public:
-    ~WorkItem() {
-        //        destroy();
-    }
-
-    WorkItem(std::function<void()> rhs) { fn = std::move(rhs); }
-
-    WorkItem(WorkItem &&rhs) { fn = std::move(rhs.fn); }
-
-public:
-    void process() { fn(); }
-    void destroy() { delete this; }
-    void drop() { delete this; }
-
-private:
-    std::function<void()> fn;
-};
-
+#include "Queue.h"
+namespace async_simple::util {
 class ThreadPool {
 public:
-    enum {
-        DEFAULT_THREADNUM = 4,
-    };
-
-    enum STOP_TYPE {
-        STOP_THREAD_ONLY = 0,
-        STOP_AND_CLEAR_QUEUE,
-        STOP_AFTER_QUEUE_EMPTY
+    struct WorkItem {
+        // Whether or not the fn is auto schedule.
+        // If the user don't assign a thread to fn,
+        // thread pool will apply random policy to fn.
+        // If enable work steal,
+        // thread pool will apply work steal policy firstly , if failed, will
+        // apply random policy to fn.
+        bool autoSchedule = false;
```

Comment:
the name of autoSchedule make me confused, you mean this item can be steal?

Reply:
If the user don't set a scheduling thread id, the autoSchedule(if it's value is true) will choose any thread  to execute, otherwise assigned the task to the given thread.

steal is another attitude, if autoSchedule is true and steal flag is true, the work steal will be trigger;
if schedule with a thread id, work steal will never trigger.

Reply:
how about to rename autoSchedule to stealable/canSteal/migratable or transferable and so on

Reply:
Right, canSteal will make the work steal syntax more clear.